File: Data Structures & Algorithms/search-2d-matrix/submission-1.js
class Solution {
    /**
     * @param {number[][]} matrix
     * @param {number} target
     * @return {boolean}
     */
    searchMatrix(matrix, target) {
        const rows = matrix.length
        const cols = matrix[0].length
        let top = 0
        let bot = rows-1
        let row = Math.floor((top+bot)/2)
        while (top <= bot) {
            row = Math.floor((top+bot)/2)
            if (target > matrix[row][cols-1]) {
                top = row +1
            }
            else if (target < matrix[row][0]) {
                bot = row -1
            }
            //we've found the right row
            else {
                break
            }
        }
        if (top > bot) return false
        let left = 0
        let right = cols -1
        while (left <= right) {
            const mid = Math.floor((left + right)/2)
            if (matrix[row][mid] > target) {
                right = mid-1
            }
            else if (matrix[row][mid] < target) {
                left = mid+1
            }
            else {
                return true
            }
        }
        return false
    }
}

File: Data Structures & Algorithms/binary-search/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums)-1
        while left <= right:
            mid = (left + right) // 2
            print(mid)
            if nums[mid] == target:
                print(f"found nums[mid]")
                return mid
            if nums[mid] < target:
                print(f"found nums[mid] < target {nums[mid]}")
                left = mid + 1
            elif nums[mid] > target:
                print(f"found nums[mid] > target {nums[mid]}")
                right = mid - 1
        return -1
        
        
